src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -180,27 +173,32 @@ public TokenData GetNextTokenData(TokenData tokenData)
         internal SyntaxToken FirstTokenOfBaseTokenLine(SyntaxToken token)
         {
             var currentTokenData = this.GetTokenData(token);
-            while (!this.IsFirstTokenOnLine(token))
+            return FirstTokenOfBaseTokenLine(ref currentTokenData);
+        }
+
+        // using ref to avoid copying the entire struct
+        private SyntaxToken FirstTokenOfBaseTokenLine(ref TokenData tokenData)
+        {
+            while (!this.IsFirstTokenOnLine(tokenData))
```

Comment:
won't a copy happen right here though?  so how does it help to avoid the copy abve?

Reply:
ok.  this is passed by `in` which can't be seen here.  either comment that on hte line, or put an `/*in*/` comment before `tokenData` to make it more clear.  Thanks!

Reply:
Added `in`